rclpy/test/test_node.py

```diff
@@ -838,6 +838,81 @@ def test_node_set_parameters_rejection(self):
+        return SetParametersResult(successful=(not any(rejected_parameters)))
+
+    def reject_parameter_callback_2(self, parameter_list):
+        rejected_parameters = (param for param in parameter_list if 'reject2' in param.name)
+        return SetParametersResult(successful=(not any(rejected_parameters)))
+
+    def reject_parameter_callback_3(self, parameter_list):
+        rejected_parameters = (param for param in parameter_list if 'reject3' in param.name)
+        return SetParametersResult(successful=(not any(rejected_parameters)))
+
+    def test_node_set_paramters_rejection_list(self):
+        # Declare a new paramters and set a list of callbacks so that it's rejected when set.
+        reject_list_parameter_tuple1 = [
+            (
+                'reject1',
+                True,
+                ParameterDescriptor()
+            ),
+            (
+                'non_reject',
+                True,
+                ParameterDescriptor()
+            ),
+            (
+                'non_reject',
+                True,
+                ParameterDescriptor()
+            )
+        ]
+
+        self.node.declare_parameters('', reject_list_parameter_tuple1)
+        self.node.add_on_set_parameters_callback(self.reject_parameter_callback_1)
+        self.node.add_on_set_parameters_callback(self.reject_parameter_callback_2)
+        self.node.add_on_set_parameters_callback(self.reject_parameter_callback_3)
+
+        result = self.node.set_parameters(
+            [
+                Parameter(
+                    name=reject_list_parameter_tuple1[0][0],
+                    value=reject_list_parameter_tuple1[0][1]
+                ),
+                Parameter(
+                    name=reject_list_parameter_tuple1[1][0],
+                    value=reject_list_parameter_tuple1[1][1]
+                ),
+                Parameter(
+                    name=reject_list_parameter_tuple1[2][0],
+                    value=reject_list_parameter_tuple1[2][1]
+                )
+            ]
+        )
+        self.assertIsInstance(result, list)
+        self.assertIsInstance(result[0], SetParametersResult)
+        self.assertFalse(result[0].successful)
+
+    def test_node_add_on_set_paramter_callback(self):
+        # Add callbacks to the list of callbacks.
+
+        callbacks = [
+            self.reject_parameter_callback,
+            self.reject_parameter_callback_1,
+            self.reject_parameter_callback_2,
+            self.reject_parameter_callback_3
+        ]
+        for callback in callbacks:
+            self.node.add_on_set_parameters_callback(callback)
+
+    def test_node_remove_from_set_callback(self):
+        # Remove callbacks from list of callbacks.
+
+        self.node.add_on_set_parameters_callback(self.reject_parameter_callback_2)
+        self.node.remove_on_set_parameters_callback(self.reject_parameter_callback_2)
```

Comment:
You should check if the callback is still there at this point, and maybe test removsl of a callback that isn't added

Reply:
@allenh1 sir but there is no api to check to get a callback ,or a method which checks if callback is present in the list..should I implement this by setting a paramter value then removing the callback that is rejecting this setting of paramter.

Reply:
@allenh1 sir is this okay?

Reply:
>  there is no api to check to get a callback ,or a method which checks if callback is present in the list

@suab321321 Since this is Python, you should be able to access them by name in the test code.

Unverified, but why not something like this?

```py
self.node.add_on_set_parameters_callback(self.reject_parameter_callback_1)
self.node.add_on_set_parameters_callback(self.reject_parameter_callback_2)
self.node.remove_on_set_parameters_callback(self.reject_parameter_callback_2)

self.assertTrue(self.reject_parameter_callback_1 in self.node._parameters_callbacks)
self.assertTrue(not (self.reject_parameter_callback_2 in self.node._parameters_callbacks))
```